### ISSUE TITLE
Replace default compiler by compiler that derives from platform

### ISSUE THREAD

**Issue by marekgluza** [csookim/qibo_cskim#27]
https://github.com/qiboteam/qibolab/issues/927

**Comment by marekgluza** [comment]
- [ ] modify the execute function in qibolab (not qibo)
- [ ] add relevant compiling info in the platform file which then will be loaded and trigger the correct compiler routines

**Comment by marekgluza** [comment]
- [ ] @scarrazza advises against environment modules https://github.com/qiboteam/nqchcluster/issues/4

**Comment by marekgluza** [comment]
For Changsoo you will need to familiarize yourself with the concept of platform files.

@sorewachigauyo can you confirm if it's accurate that platform runcards:
- they include channels and readout mappings, that's in line with the chip ports physically
- then there is qibocal which can automatically update runcard files and e.g. fill in the numerical values of for frequencies of qubits etc obtained by fitting during calibration
- we need to design a systematic way to list types of compilers (for now we can use 'default')
- then the platform file will have a compiler type which should specify which gates are the primitive gates; these gates should be the gates callibrated by qibocal
- then qibolab Circuit.execute function should pull the type of compiler from the runcard and then express everything into the primitive gates specified there

**Comment by sorewachigauyo** [comment]
Yes

1. qibolab describes the channel mapping for the pulses to target the physical ports
2. Qibocal has the routines to fill in the parameters for the runcard that is needed for higher level tasks, such as other routines and circuit execution
3. Sure
4. The native gates are located in the parameter file (aka https://github.com/qiboteam/qibolab_platforms_nqch/blob/main/icarusq_iqm5q/parameters.json, RX, CNOT/CZ)
5. There are already some compiler rules to map some gates to pulses, like the GPI2, in Qibolab.

**Comment by scarrazza** [comment]
Exactly, I suggest to discuss with Stavros, Alessandro and Hayk about 3.

Concerning the environmnet for slurm, I would prefer to have just 2 possibilities: qibo/stable and qibo/latest. Extra configurations (platform specific, such as `QIBO_BACKEND` and `QIBO_PLATFORM`) should go the slurm prolog script.

**Comment by alecandido** [comment]
Hopefully, there won't be much need for environment variables in any case. Since, usually, you want to target a specific chip, the information about the platform can be contained within the Qibocal runcard (or, better, the Qibocal script).

About the natives, things might change in the future, since we may design better ways to support different architectures. But right now (and for a while) it is exactly as @sorewachigauyo wrote, but you should not access them directly in the `parameters.json` file, but through the platform (the `Qubit` and `QubitPair` objects track their natives).

The compiler is handled by the platform, and for the time being it would be advisable to be set in there (though, if we start having multiple predefined compilers, we might distribute them separately from platforms).
The transpiler instead is pretty much platform independent, and should only depend on the chip connectivity. I was planning to orchestrate it at the level of the `GlobalBackend`, making the connectivity a standard feature for Qibo backends (i.e. even in simulation, and it will default to full connectivity). However, it is not there yet, so any solution you find is fine for the time being. At most, I will propose a further transition later on.

**Comment by alecandido** [comment]
> I assume that sooner or later qibo will have to enable a default transpiler (which might be the same for all platforms or not) when allocating the qibolab backend.

Yes, that's correct. And the idea is to have a transpiler around even in simulation. By default, it won't do anything (but you may want to specify a connectivity and non-trivial transpiler even in simulation).

> May I suggest to move this part of the discussion to qibolab/qibo?

Even fine.
I haven't done it yet because I thought it premature, since I was not going to work on it. But we're also opening issues for the long term, and maybe someone else will do it first. So that's definitely correct.
